### PR TITLE
Update kubeadm token to work as expected

### PR DESCRIPTION
**Pull request by luxas**
**What this PR does / why we need it**:

Follows up: https://github.com/kubernetes/kubernetes/pull/41509

Updates `kubeadm token` to work as discussed in https://docs.google.com/document/d/1deJYPIF4LmhGjDVaqrswErIrV7mtwJgovtLnPCDxP7U/edit#
Promotes the command from the `ex` subcommand which now is named `alpha` for clarity. (This will later become `kubeadm alpha phase`)

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

Example UX:
```console
sudo ./kubeadm token --help
This command will manage Bootstrap Token for you.
  Please note this usage of this command is optional, and mostly for advanced users.

In short, Bootstrap Tokens are used for establishing bidirectional trust between a client and a server.
A Bootstrap Token can be used when a client (for example a node that's about to join the cluster) needs
to trust the server it is talking to. Then a Bootstrap Token with the "signing" usage can be used.
Bootstrap Tokens can also function as a way to allow short-lived authentication to the API Server
(the token serves as a way for the API Server to trust the client), for example for doing the TLS Bootstrap.

What is a Bootstrap Token more exactly?
 - It is a Secret in the kube-system namespace of type "bootstrap.kubernetes.io/token".
 - A Bootstrap Token must be of the form "[a-z0-9]{6}.[a-z0-9]{16}"; the former part is the public Token ID,
   and the latter is the Token Secret, which must be kept private at all circumstances.
 - The name of the Secret must be named "bootstrap-token-(token-id)".

You can read more about Bootstrap Tokens in this proposal:

  https://github.com/kubernetes/community/blob/master/contributors/design-proposals/bootstrap-discovery.md

Usage:
  kubeadm token [flags]
  kubeadm token [command]

Available Commands:
  create      Create bootstrap tokens on the server.
  delete      Delete bootstrap tokens on the server.
  generate    Generate and print a bootstrap token, but do not create it on the server.
  list        List bootstrap tokens on the server.

Flags:
      --kubeconfig string   The KubeConfig file to use for talking to the cluster (default "/etc/kubernetes/admin.conf")

Use "kubeadm token [command] --help" for more information about a command.
lucas@THENINJA:~/luxas/kubernetes$ sudo ./kubeadm token list
TOKEN                     TTL         EXPIRES   USAGES                   DESCRIPTION
70c388.41a07b703aa4bedf   <forever>   <never>   authentication,signing   The default bootstrap token generated by 'kubeadm init'.
lucas@THENINJA:~/luxas/kubernetes$ sudo ./kubeadm token create
c57e6a.abb75fa1debe555f
lucas@THENINJA:~/luxas/kubernetes$ sudo ./kubeadm token list
TOKEN                     TTL         EXPIRES   USAGES                   DESCRIPTION
70c388.41a07b703aa4bedf   <forever>   <never>   authentication,signing   The default bootstrap token generated by 'kubeadm init'.
c57e6a.abb75fa1debe555f   <forever>   <never>   authentication,signing   <none>
lucas@THENINJA:~/luxas/kubernetes$ sudo ./kubeadm token create s
token ["s"] was not of form ["^([a-z0-9]{6})\\.([a-z0-9]{16})$"]
lucas@THENINJA:~/luxas/kubernetes$ sudo ./kubeadm token create c57e6a.abb75fa1debe555f
a token with id "c57e6a" already exists
lucas@THENINJA:~/luxas/kubernetes$ sudo ./kubeadm token delete c57e6a.abb75fa1debe555f
bootstrap token with id "c57e6a" deleted
```

**Release note**:

```release-note
NONE
```
@dmmcquay @jbeda @mikedanese @errordeveloper @pires

### CODE REVIEW

File: cmd/kubeadm/app/cmd/token.go [hunk 1]
```diff
@@ -85,20 +93,26 @@ func NewCmdToken(out io.Writer, errW io.Writer) *cobra.Command {
...
+		Use:   "delete [token]",
 		Short: "Delete bootstrap tokens on the server.",
 		Run: func(tokenCmd *cobra.Command, args []string) {
 			if len(args) < 1 {
-				kubeadmutil.CheckErr(fmt.Errorf("missing subcommand; 'token delete' is missing token of form [\"^([a-z0-9]{6})$\"]"))
+				kubeadmutil.CheckErr(fmt.Errorf("missing subcommand; 'token delete' is missing token of form [%q]", tokenutil.TokenIDRegexpString))
 			}
-			err := RunDeleteToken(out, tokenCmd, args[0])
+			client, err := kubemaster.CreateClientFromFile(getKubeConfigPath())
+			kubeadmutil.CheckErr(err)
```
Comment (dmmcquay) [inline cmd/kubeadm/app/cmd/token.go]: same

File: cmd/kubeadm/app/cmd/token.go [hunk 2]
```diff
@@ -130,24 +148,26 @@ func NewCmdTokenGenerate(out io.Writer) *cobra.Command {
...
 // RunCreateToken generates a new bootstrap token and stores it as a secret on the server.
-func RunCreateToken(out io.Writer, cmd *cobra.Command, tokenDuration time.Duration, token string) error {
-	client, err := kubemaster.CreateClientFromFile(path.Join(kubeadmapi.GlobalEnvParams.KubernetesDir, kubeconfig.AdminKubeConfigFileName))
-	if err != nil {
-		return err
-	}
+func RunCreateToken(out io.Writer, client *clientset.Clientset, token string, tokenDuration time.Duration, usages []string) error {
 
-	parsedID, parsedSecret, err := tokenutil.ParseToken(token)
+	td := &kubeadmapi.TokenDiscovery{}
```
Comment (dmmcquay) [inline cmd/kubeadm/app/cmd/token.go]: Why don't you try to create the token using `td.ID, td.Secret, err = tokenutil.ParseToken(token)`, and if it returns an err, you then try to create the token using generate.  Too many checks, you could make it read more clearly.

File: cmd/kubeadm/app/cmd/token.go [hunk 3]
```diff
@@ -178,61 +194,86 @@ func RunListTokens(out io.Writer, errW io.Writer, cmd *cobra.Command) error {
...
 		}
-		fmt.Fprintf(w, "%s\t%s\t%s\n", tokenId, tokenutil.BearerToken(td), expires)
+
+		usages := []string{}
+		for k, v := range secret.Data {
+			// Skip all fields that don't include this prefix
+			if !strings.Contains(k, bootstrapapi.BootstrapTokenUsagePrefix) {
+				continue
+			}
+			// Skip those that hasn't this usage set to true
```
Comment (dmmcquay) [inline cmd/kubeadm/app/cmd/token.go]: Edit comment "Skip those that don't have this usage set to true"

File: cmd/kubeadm/app/cmd/token.go [hunk 4]
```diff
@@ -178,61 +194,86 @@ func RunListTokens(out io.Writer, errW io.Writer, cmd *cobra.Command) error {
...
 
-	tokenSecretName := fmt.Sprintf("%s%s", kubeadmconstants.BootstrapTokenSecretPrefix, tokenId)
-	if err := client.Secrets(metav1.NamespaceSystem).Delete(tokenSecretName, nil); err != nil {
+	tokenSecretName := kubeadmconstants.BootstrapTokenSecretPrefix + tokenId
+	if err := client.CoreV1().Secrets(metav1.NamespaceSystem).Delete(tokenSecretName, nil); err != nil {
 		return fmt.Errorf("failed to delete bootstrap token [%v]", err)
 	}
-	fmt.Fprintf(out, "[token] bootstrap token deleted: %s\n", tokenId)
-
+	fmt.Fprintf(out, "bootstrap token with id %q deleted\n", tokenId)
```
Comment (dmmcquay) [inline cmd/kubeadm/app/cmd/token.go]: are we moving away from having [foo] labels?

File: cmd/kubeadm/app/cmd/token.go [hunk 5]
```diff
@@ -61,22 +64,27 @@ func NewCmdToken(out io.Writer, errW io.Writer) *cobra.Command {
...
 		},
 	}
 	createCmd.PersistentFlags().DurationVar(&tokenDuration,
 		"ttl", kubeadmconstants.DefaultTokenDuration, "The duration before the token is automatically deleted. 0 means 'never expires'.")
-	createCmd.PersistentFlags().StringVar(
-		&token, "token", "",
-		"Shared secret used to secure cluster bootstrap. If none is provided, one will be generated for you.",
-	)
+	createCmd.PersistentFlags().StringSliceVar(&usages,
+		"usages", kubeadmconstants.DefaultTokenUsages, "The functions this token can be used for.")
```
Comment (dmmcquay) [inline cmd/kubeadm/app/cmd/token.go]: I don't like the word 'usages' here. Sounds strange. Anyone agree?

File: cmd/kubeadm/app/cmd/token.go [hunk 6]
```diff
@@ -178,61 +194,86 @@ func RunListTokens(out io.Writer, errW io.Writer, cmd *cobra.Command) error {
...
-		secretExpiration, ok := secret.Data[bootstrapapi.BootstrapTokenExpirationKey]
-		if ok {
-			expireTime, err := time.Parse(time.RFC3339, string(secretExpiration))
+		secretExpiration := getSecretString(&secret, bootstrapapi.BootstrapTokenExpirationKey)
+		if len(secretExpiration) > 0 {
+			expireTime, err := time.Parse(time.RFC3339, secretExpiration)
 			if err != nil {
-				return fmt.Errorf("error parsing expiration time [%v]", err)
+				fmt.Fprintf(errW, "can't parse expiration time of bootstrap token %s\n", secret.Name)
+				continue
```
Comment (dmmcquay) [inline cmd/kubeadm/app/cmd/token.go]: Why don't we return an error if the time parsing fails? It seems like this function never returns an error that isn't nil inside the iteration of Secret Items.

File: cmd/kubeadm/app/util/token/tokens.go [hunk 7]
```diff
@@ -32,10 +32,10 @@ const (
 )
 
 var (
-	tokenIDRegexpString = "^([a-z0-9]{6})$"
-	tokenIDRegexp       = regexp.MustCompile(tokenIDRegexpString)
-	tokenRegexpString   = "^([a-z0-9]{6})\\:([a-z0-9]{16})$"
-	tokenRegexp         = regexp.MustCompile(tokenRegexpString)
+	TokenIDRegexpString = "^([a-z0-9]{6})$"
+	tokenIDRegexp       = regexp.MustCompile(TokenIDRegexpString)
+	TokenRegexpString   = "^([a-z0-9]{6})\\.([a-z0-9]{16})$"
```
Comment (dmmcquay) [inline cmd/kubeadm/app/util/token/tokens.go]: You need to update the tests. They fail because they were expecting a ':'

**Review by dmmcquay — CHANGES_REQUESTED** [review]
Need to update tests and some minor comments.

File: cmd/kubeadm/app/phases/token/bootstrap.go [hunk 8]
```diff
@@ -36,18 +36,18 @@ const (
...
 // UpdateOrCreateToken attempts to update a token with the given ID, or create if it does
 // not already exist.
-func UpdateOrCreateToken(client *clientset.Clientset, d *kubeadmapi.TokenDiscovery, tokenDuration time.Duration) error {
+func UpdateOrCreateToken(client *clientset.Clientset, d *kubeadmapi.TokenDiscovery, tokenDuration time.Duration, usages []string) error {
 	// Let's make sure the token is valid
 	if valid, err := tokenutil.ValidateToken(d); !valid {
 		return err
 	}
-	secretName := fmt.Sprintf("%s%s", kubeadmconstants.BootstrapTokenSecretPrefix, d.ID)
+	secretName := kubeadmconstants.BootstrapTokenSecretPrefix + d.ID
```
Comment (dmmcquay) [inline cmd/kubeadm/app/phases/token/bootstrap.go]: I prefer fmt.Sprintf, it scales better with codebase.

**Review by dmmcquay — CHANGES_REQUESTED** [review]

File: cmd/kubeadm/app/cmd/token.go (hunk 5 above)
Comment (jbeda) [inline cmd/kubeadm/app/cmd/token.go]: It builds on the terminology used for x509 certs -- https://en.wikipedia.org/wiki/X.509#Extensions_informing_a_specific_usage_of_a_certificate.

Comment (dmmcquay) [inline cmd/kubeadm/app/cmd/token.go]: fair enough! I just thought the plural of it sounded strange. But, if there is support, totally fine.

File: cmd/kubeadm/app/cmd/token.go [hunk 9]
```diff
@@ -85,20 +93,26 @@ func NewCmdToken(out io.Writer, errW io.Writer) *cobra.Command {
 		Use:   "list",
 		Short: "List bootstrap tokens on the server.",
 		Run: func(tokenCmd *cobra.Command, args []string) {
-			err := RunListTokens(out, errW, tokenCmd)
+			client, err := kubemaster.CreateClientFromFile(getKubeConfigPath())
+			kubeadmutil.CheckErr(err)
```
Comment (jbeda) [inline cmd/kubeadm/app/cmd/token.go]: Regardless, it is a common pattern in many of the utilities in k8s.  I see this in `kubectl` and `kubefed` also.  If we want to change it it should probably be done in a different PR.

File: cmd/kubeadm/app/cmd/token.go (hunk 2 above)
Comment (jbeda) [inline cmd/kubeadm/app/cmd/token.go]: If someone passes in a malformed token we should error explicitly.  If we instead ignore that and generate a random token they may think that everything worked until they try and used their malformed token elsewhere.

I'd rather fail early with clear message if the user passes a malformed token.

File: cmd/kubeadm/app/cmd/token.go (hunk 6 above)
Comment (jbeda) [inline cmd/kubeadm/app/cmd/token.go]: IMO, if one secret is malformed we want to be able to print out the rest of them.  If we return an error here then one bad secret spoils everything.

File: cmd/kubeadm/app/cmd/token.go [hunk 10]
```diff
@@ -178,61 +194,86 @@ func RunListTokens(out io.Writer, errW io.Writer, cmd *cobra.Command) error {
...
+	tokenId := tokenIdOrToken
+	if err := tokenutil.ParseTokenID(tokenIdOrToken); err != nil {
+		if tokenId, _, err = tokenutil.ParseToken(tokenIdOrToken); err != nil {
+			return fmt.Errorf("given token or token id %q didn't match pattern [%q] or [%q]", tokenIdOrToken, tokenutil.TokenIDRegexpString, tokenutil.TokenRegexpString)
+		}
 	}
 
-	tokenSecretName := fmt.Sprintf("%s%s", kubeadmconstants.BootstrapTokenSecretPrefix, tokenId)
-	if err := client.Secrets(metav1.NamespaceSystem).Delete(tokenSecretName, nil); err != nil {
+	tokenSecretName := kubeadmconstants.BootstrapTokenSecretPrefix + tokenId
```
Comment (jbeda) [inline cmd/kubeadm/app/cmd/token.go]: I'd rather not take a dependency on the name and instead iterate over the secrets finding the token.  The name of the secret is really just a suggestion.  If someone uses another tool to create secrets we may end up in a situation where a secret shows up with `kubeadm token list` but `kubeadm token delete` won't delete it.  That would be frustrating and confusing.

File: cmd/kubeadm/app/phases/token/bootstrap.go (hunk 8 above)
Comment (jbeda) [inline cmd/kubeadm/app/phases/token/bootstrap.go]: Same thing here -- we should really make this look at all secrets vs making assumptions about name.

**Review by jbeda — CHANGES_REQUESTED** [review]
Looks pretty good to me.  The biggest problem I have is that we are making assumptions on the secret name.  The spec currently says that name is suggested but not required.

The signer and authenticator don't make this assumption.  In addition, `kubeadm token list` doesn't either.

So -- we should do one of the following:
1) Check this in as-is and have a bug/TODO to fix it.  Let's not let that go too long.
2) Fix it now.
3) Redefine it so that the name is required.  We'd probably want to make the server side code validate things too.  And have `kubeadm token list` warn if the name doesn't match the contents.

I prefer 2 but could live with 1.  3 seems wonky.

File: cmd/kubeadm/app/cmd/token.go (hunk 2 above)
Comment (dmmcquay) [inline cmd/kubeadm/app/cmd/token.go]: I don't mean to not error out, I just didn't like the logic of the code. Just use the error checking built into the function calls.

**Comment by pires** [comment]
The only question I have is about dropping `:` for `.` as this was causing problems when parsing URLs. Also, the tests don't pass so I'm sure you are not running them on your side and you should.

**Comment by jbeda** [comment]
@pires -- where was the token showing up in URLs?  Sensitive secrets should generally not show up in a URL as they have a history of being cached and exposed.  Or was this with the fake-o `token://` URL?  On going talk of kubeadm UX in this area is happening in https://github.com/kubernetes/community/pull/381.

**Comment by luxas** [comment]
@jbeda It was just a note that what we're proposing in https://github.com/kubernetes/community/pull/381 isn't implemented for `kubeadm join` yet. 
But @dmmcquay is on that. He will come up with a PR for that soon I think.

@pires Of course I'm running tests, and I'm implementing new ones as well.
On the other hand I'm sorry for not marking this WIP during the time we discussed the details in  https://github.com/kubernetes/community/pull/381. Will do so now. Soon, a new changeset of this PR will be up with all the latest and greatest content.

**Comment by pires** [comment]
@jbeda `kubeadm join --discovery token://<token>@node`.

File: cmd/kubeadm/app/cmd/token.go (hunk 10 above)
Comment (jbeda) [inline cmd/kubeadm/app/cmd/token.go]: This is stale now as we decided to make the name required.

**Review by jbeda — APPROVED** [review]
This LGTM.  Only real problem is where we pull kubeconfig from.  We will want to allow `kubeadm token` to run against a remote cluster.  But that can be another PR.

**Comment by luxas** [comment]
@jbeda Rebased and updated to implement the latest design decisions in https://github.com/kubernetes/community/pull/381

Even though a single test or something might fail -- please give your lgtm if possible and I'll self-apply later

(Jenkins is overloaded atm, so we don't have a clue about tests currently)

**Comment by luxas** [comment]
@k8s-bot test this

**Comment by luxas** [comment]
I'll rebase as soon as https://github.com/kubernetes/kubernetes/pull/41857 merges.

@jbeda PTAL and please /lgtm, I'll take care of rebasing tomorrow, but the faster you can lgtm the better

File: cmd/kubeadm/app/cmd/token.go [hunk 11]
```diff
@@ -60,22 +63,33 @@ func NewCmdToken(out io.Writer, errW io.Writer) *cobra.Command {
...
+	createCmd.Flags().DurationVar(&tokenDuration,
 		"ttl", kubeadmconstants.DefaultTokenDuration, "The duration before the token is automatically deleted. 0 means 'never expires'.")
-	createCmd.PersistentFlags().StringVar(
-		&token, "token", "",
-		"Shared secret used to secure cluster bootstrap. If none is provided, one will be generated for you.",
-	)
+	createCmd.Flags().StringSliceVar(&usages,
+		"usages", kubeadmconstants.DefaultTokenUsages, "The functions this token can be used for.")
+	createCmd.Flags().StringVar(&description,
+		"description", "", "A human-friendly description what this token is used for.")
```
Comment (jbeda) [inline cmd/kubeadm/app/cmd/token.go]: reword? "A human friendly description of how this token is used."

File: cmd/kubeadm/app/cmd/token.go [hunk 12]
```diff
@@ -60,22 +63,33 @@ func NewCmdToken(out io.Writer, errW io.Writer) *cobra.Command {
...
 	}
-	createCmd.PersistentFlags().DurationVar(&tokenDuration,
+	createCmd.Flags().DurationVar(&tokenDuration,
 		"ttl", kubeadmconstants.DefaultTokenDuration, "The duration before the token is automatically deleted. 0 means 'never expires'.")
-	createCmd.PersistentFlags().StringVar(
-		&token, "token", "",
-		"Shared secret used to secure cluster bootstrap. If none is provided, one will be generated for you.",
-	)
+	createCmd.Flags().StringSliceVar(&usages,
+		"usages", kubeadmconstants.DefaultTokenUsages, "The functions this token can be used for.")
```
Comment (jbeda) [inline cmd/kubeadm/app/cmd/token.go]: reword? "functions" sounds like programing language function. "The ways in which this token can be used"

File: cmd/kubeadm/app/cmd/token.go [hunk 13]
```diff
@@ -60,22 +63,33 @@ func NewCmdToken(out io.Writer, errW io.Writer) *cobra.Command {
...
-	var token string
+	tokenCmd.PersistentFlags().StringVar(&kubeConfigFile,
+		"kubeconfig", "/etc/kubernetes/admin.conf", "The KubeConfig file to use for talking to the cluster")
+
+	var usages []string
 	var tokenDuration time.Duration
+	var description string
 	createCmd := &cobra.Command{
-		Use:   "create",
+		Use:   "create [token]",
```
Comment (jbeda) [inline cmd/kubeadm/app/cmd/token.go]: The `[token]` parameter is undefined in the help text.  I'd do `create [token-value]` and then change the description to document that "The token-value is the actual token to write.  This should be a securely generated random token of the form [a-z0-9]{6}.[a-z0-9]{16}."

File: cmd/kubeadm/app/cmd/token.go [hunk 14]
```diff
@@ -84,20 +98,26 @@ func NewCmdToken(out io.Writer, errW io.Writer) *cobra.Command {
...
+
+			err = RunListTokens(out, errW, client)
 			kubeadmutil.CheckErr(err)
 		},
 	}
 	tokenCmd.AddCommand(listCmd)
 
 	deleteCmd := &cobra.Command{
-		Use:   "delete",
+		Use:   "delete [token]",
```
Comment (jbeda) [inline cmd/kubeadm/app/cmd/token.go]: nit: Same comment on token param here.

File: cmd/kubeadm/app/cmd/token.go (hunk 4 above)
Comment (luxas) [inline cmd/kubeadm/app/cmd/token.go]: not generally, but it doesn't make sense to have it for only one line

File: cmd/kubeadm/app/cmd/token.go [hunk 15]
```diff
@@ -177,61 +196,93 @@ func RunListTokens(out io.Writer, errW io.Writer, cmd *cobra.Command) error {
...
 	}
 
 	w := tabwriter.NewWriter(out, 10, 4, 3, ' ', 0)
-	fmt.Fprintln(w, "ID\tTOKEN\tTTL")
-	for _, secret := range results.Items {
-		tokenId, ok := secret.Data[bootstrapapi.BootstrapTokenIDKey]
-		if !ok {
-			fmt.Fprintf(errW, "[token] bootstrap token has no token-id data: %s\n", secret.Name)
+	fmt.Fprintln(w, "TOKEN\tTTL\tEXPIRES\tUSAGES\tDESCRIPTION")
+	for _, secret := range secrets.Items {
```
Comment (jbeda) [inline cmd/kubeadm/app/cmd/token.go]: nit+: We should skip/ignore/warn on tokens that don't match our naming requirement.

**Review by jbeda — APPROVED** [review]
Mostly nits.  LGTM

**Review by dmmcquay — APPROVED** [review]
/lgtm

**Comment by luxas** [comment]
@k8s-bot non-cri e2e test this

**Comment by luxas** [comment]
@k8s-bot cvm gce e2e test this

**Comment by dmmcquay** [comment]
@k8s-bot non-cri e2e test this

**Comment by alexellis** [comment]
@luxas is there a way to generate the whole join command? I feel like I have to reset the entire cluster to be able to "reach" the magic string i.e.

```
  kubeadm join --token TOK MASTER_IP:6443 --discovery-token-ca-cert-hash sha256:HASH
```

This is very easy with Swarm - `docker swarm join-token worker`... would be great if you could document or provide a way to do this with `kubeadm`.

**Comment by luxas** [comment]
@alexellis Please open a new issue for this and cc @mattmoyer and @jamiehannaford please.
TL;DR; the token can be fetched from `kubeadm token list`, the CA hash with `openssl` (as documented) and the master IP from the cluster.

Room for improvement indeed.

**Comment by 1605200517** [comment]
https://github.com/kubernetes/kubernetes/issues/55108 @luxas @alexellis
